pwndbg/gdblib/events.py

```diff
@@ -75,6 +77,69 @@ def on_stop(self) -> None:
 gdb.events.start = StartEvent()
 
 
+def _is_safe_event_packet():
+    try:
+        gdb.selected_frame()
+    except gdb.error as e:
+        if "Remote 'g' packet reply is too long" in str(e):
+            return False
+    return True
+
+
+def _is_safe_event_thread():
+    try:
+        gdb.newest_frame()
+    except gdb.error as e:
+        if "Selected thread is running" in str(e):
+            return False
+    return True
+
+
+class _DelayedEventHandler:
+    def __init__(self, func: Callable[[], Any]):
+        self.func = func
+
+    def __call__(self):
+        self.func()
+
+
+def wrap_safe_event_handler(event_handler: Callable[P, T]) -> Callable[P, T]:
+    """
+    Wraps an event handler to ensure it is only executed when the event is safe.
+    Invalid events are queued and executed later when safe.
+
+    Note: Avoid using `gdb.post_event` because of another bug in gdbserver
+    where the `gdb.newest_frame` function may not work properly.
+
+    Workaround to fix bug in gdbserver: https://github.com/pwndbg/pwndbg/issues/2576
+    """
+    queued_invalid_events: Deque[Callable[..., Any]] = deque()
+
+    def _loop_until_thread_ok():
+        if not queued_invalid_events:
+            return
+
+        if not _is_safe_event_thread():
+            gdb.post_event(_DelayedEventHandler(_loop_until_thread_ok))
```

Comment:
Shouldn't this be `gdb.post_event(event_handler)`?

Also the `_DelayedEventHandler` class is redundant - it does not add anything to the code iiuc.